// angular-imask/bundles/angular-imask.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('imask'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-imask', ['exports', '@angular/core', '@angular/forms', 'imask', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['angular-imask'] = {}, global.ng.core, global.ng.forms, global.IMask, global.ng.common));
}(this, (function (exports, i0, forms, IMask, common) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var IMask__default = /*#__PURE__*/_interopDefaultLegacy(IMask);

    var IMaskFactory = /** @class */ (function () {
        function IMaskFactory() {
        }
        return IMaskFactory;
    }());
    IMaskFactory.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function IMaskFactory_Factory() { return new IMaskFactory(); }, token: IMaskFactory, providedIn: "root" });
    IMaskFactory.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];

    var MASKEDINPUT_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: i0.forwardRef(function () { return IMaskDirective; }),
        multi: true
    };
    var DEFAULT_IMASK_ELEMENT = function (elementRef) { return elementRef.nativeElement; };
    var ɵ0 = DEFAULT_IMASK_ELEMENT;
    var IMaskDirective = /** @class */ (function () {
        function IMaskDirective(_elementRef, _renderer, _factory, _compositionMode) {
            this._elementRef = _elementRef;
            this._renderer = _renderer;
            this._factory = _factory;
            this._compositionMode = _compositionMode;
            // init here to support AOT (TODO may be will work with ng-packgr - need to check)
            this.onTouched = function () { };
            this.onChange = function () { };
            this.imaskElement = DEFAULT_IMASK_ELEMENT;
            this.accept = new i0.EventEmitter();
            this.complete = new i0.EventEmitter();
            this._viewInitialized = false;
            this._composing = false;
            this._writing = false;
            if (this._compositionMode == null) {
                this._compositionMode = !this._isAndroid();
            }
        }
        Object.defineProperty(IMaskDirective.prototype, "element", {
            get: function () {
                return this.imaskElement(this._elementRef, this);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(IMaskDirective.prototype, "maskValue", {
            get: function () {
                if (!this.maskRef)
                    return this.element.value;
                if (this.unmask === 'typed')
                    return this.maskRef.typedValue;
                if (this.unmask)
                    return this.maskRef.unmaskedValue;
                return this.maskRef.value;
            },
            set: function (value) {
                if (this.maskRef) {
                    if (this.unmask === 'typed')
                        this.maskRef.typedValue = value;
                    else if (this.unmask)
                        this.maskRef.unmaskedValue = value;
                    else
                        this.maskRef.value = value;
                }
                else {
                    this._renderer.setProperty(this.element, 'value', value);
                }
            },
            enumerable: false,
            configurable: true
        });
        IMaskDirective.prototype.ngAfterViewInit = function () {
            if (this.imask)
                this.initMask();
            this._viewInitialized = true;
        };
        IMaskDirective.prototype.ngOnChanges = function (changes) {
            if (changes.elementRef && !this.imaskElement)
                this.imaskElement = DEFAULT_IMASK_ELEMENT;
            if (!changes.imask || !this._viewInitialized)
                return;
            if (this.imask) {
                if (this.maskRef)
                    this.maskRef.updateOptions(this.imask);
                else {
                    this.initMask();
                    this.onChange(this.maskValue);
                }
            }
            else {
                this.destroyMask();
            }
        };
        IMaskDirective.prototype.destroyMask = function () {
            if (this.maskRef) {
                this.maskRef.destroy();
                delete this.maskRef;
            }
        };
        IMaskDirective.prototype.ngOnDestroy = function () {
            this.destroyMask();
            this.accept.complete();
            this.complete.complete();
        };
        IMaskDirective.prototype.beginWrite = function (value) {
            this._writing = true;
            this._writingValue = value;
        };
        IMaskDirective.prototype.endWrite = function () {
            this._writing = false;
            return this._writingValue;
        };
        IMaskDirective.prototype.writeValue = function (value) {
            value = value == null ? '' : value;
            if (this.maskRef) {
                this.beginWrite(value);
                if (this.maskValue !== value ||
                    // handle cases like Number('') === 0,
                    // for details see https://github.com/uNmAnNeR/imaskjs/issues/134
                    (typeof value !== 'string' && this.maskRef.value === '') &&
                        !this.maskRef.el.isActive) {
                    this.maskValue = value;
                }
            }
            else {
                this._renderer.setProperty(this.element, 'value', value);
            }
        };
        IMaskDirective.prototype._onAccept = function () {
            var value = this.maskValue;
            // if value was not changed during writing don't fire events
            // for details see https://github.com/uNmAnNeR/imaskjs/issues/136
            if (this._writing && value === this.endWrite())
                return;
            this.onChange(value);
            this.accept.emit(value);
        };
        IMaskDirective.prototype._onComplete = function () {
            this.complete.emit(this.maskValue);
        };
        IMaskDirective.prototype.initMask = function () {
            this.maskRef = this._factory.create(this.element, this.imask)
                .on('accept', this._onAccept.bind(this))
                .on('complete', this._onComplete.bind(this));
        };
        IMaskDirective.prototype.setDisabledState = function (isDisabled) {
            this._renderer.setProperty(this.element, 'disabled', isDisabled);
        };
        IMaskDirective.prototype.registerOnChange = function (fn) { this.onChange = fn; };
        IMaskDirective.prototype.registerOnTouched = function (fn) { this.onTouched = fn; };
        IMaskDirective.prototype._handleInput = function (value) {
            // if mask is attached all input goes throw mask
            if (this.maskRef)
                return;
            if (!this._compositionMode || (this._compositionMode && !this._composing)) {
                this.onChange(value);
            }
        };
        IMaskDirective.prototype._compositionStart = function () { this._composing = true; };
        IMaskDirective.prototype._compositionEnd = function (value) {
            this._composing = false;
            this._compositionMode && this._handleInput(value);
        };
        IMaskDirective.prototype._isAndroid = function () {
            return /android (\d+)/.test(navigator.userAgent.toLowerCase());
        };
        return IMaskDirective;
    }());
    IMaskDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[imask]',
                    exportAs: 'imask',
                    host: {
                        '(input)': '_handleInput($event.target.value)',
                        '(blur)': 'onTouched()',
                        '(compositionstart)': '_compositionStart()',
                        '(compositionend)': '_compositionEnd($event.target.value)'
                    },
                    providers: [MASKEDINPUT_VALUE_ACCESSOR],
                },] }
    ];
    IMaskDirective.ctorParameters = function () { return [
        { type: i0.ElementRef },
        { type: i0.Renderer2 },
        { type: IMaskFactory },
        { type: Boolean, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [forms.COMPOSITION_BUFFER_MODE,] }] }
    ]; };
    IMaskDirective.propDecorators = {
        imask: [{ type: i0.Input }],
        unmask: [{ type: i0.Input }],
        imaskElement: [{ type: i0.Input }],
        accept: [{ type: i0.Output }],
        complete: [{ type: i0.Output }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /*
     * Transforms value through mask
     * Takes mask and optionally `from` and `to` pipe types.
     * Usage:
     *   value | imask:MASK_OR_MASKED:opt_from:opt_to
     * Example:
     *   {{ 2 | imask:mask }}
    */
    var IMaskPipe = /** @class */ (function () {
        function IMaskPipe() {
        }
        IMaskPipe.prototype.transform = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return IMask.pipe.apply(void 0, __spreadArray([], __read(args)));
        };
        return IMaskPipe;
    }());
    IMaskPipe.decorators = [
        { type: i0.Pipe, args: [{ name: 'imask' },] }
    ];

    var DefaultImaskFactory = /** @class */ (function () {
        function DefaultImaskFactory() {
        }
        DefaultImaskFactory.prototype.create = function (el, opts) {
            return IMask__default['default'](el, opts);
        };
        return DefaultImaskFactory;
    }());
    DefaultImaskFactory.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function DefaultImaskFactory_Factory() { return new DefaultImaskFactory(); }, token: DefaultImaskFactory, providedIn: "root" });
    DefaultImaskFactory.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];

    var IMaskDirectiveModule = /** @class */ (function () {
        function IMaskDirectiveModule() {
        }
        return IMaskDirectiveModule;
    }());
    IMaskDirectiveModule.decorators = [
        { type: i0.NgModule, args: [{
                    imports: [common.CommonModule],
                    declarations: [IMaskDirective],
                    providers: [{ provide: IMaskFactory, useClass: DefaultImaskFactory }],
                    exports: [IMaskDirective]
                },] }
    ];

    var IMaskModule = /** @class */ (function () {
        function IMaskModule() {
        }
        return IMaskModule;
    }());
    IMaskModule.decorators = [
        { type: i0.NgModule, args: [{
                    imports: [common.CommonModule, IMaskDirectiveModule],
                    declarations: [IMaskPipe],
                    exports: [IMaskPipe, IMaskDirectiveModule]
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    Object.defineProperty(exports, 'PIPE_TYPE', {
        enumerable: true,
        get: function () {
            return IMask.PIPE_TYPE;
        }
    });
    Object.defineProperty(exports, 'pipe', {
        enumerable: true,
        get: function () {
            return IMask.pipe;
        }
    });
    exports.IMaskDirective = IMaskDirective;
    exports.IMaskDirectiveModule = IMaskDirectiveModule;
    exports.IMaskFactory = IMaskFactory;
    exports.IMaskModule = IMaskModule;
    exports.IMaskPipe = IMaskPipe;
    exports.MASKEDINPUT_VALUE_ACCESSOR = MASKEDINPUT_VALUE_ACCESSOR;
    exports.ɵ0 = ɵ0;
    exports.ɵa = DefaultImaskFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
